Add Optional "Debug-Mode" to `beatBruteBot.py`

FedeCuci opened the pull request:
Hey @Celqaz I have edited your automation script so that users have the choice to "debug" maps that their bot is not able to complete. This is done by pressing `ENTER` for each move, so that the user has more control over how they check how their bot behaves. If a user is stuck and is done debugging a certain map, any key can be pressed to start debugging the next map.

Also, every position that the bot has been in is marked with a number, which signifies the number of times the bot has been in that particular cell. This is only done until the number 9, as double-digit numbers would alter the outline of the map. Therefore, if the bot has visited a cell more than nine times, the cell is replaced with an "S", which stands for "Stuck".

Although a minor change, I found that being able to debug my bot following every iteration has helped me a lot. The changes do not force the user to debug, but rather give them a choice so that it is not too intrusive. Would you mind checking the new steps I have added in the Readme.md and letting me know if you are able to reproduce them? Also, let me know if the script works fine with your bot too! If so, I think this would be a nice addition.

For the future, it might be best to just include the new `Game.py` and `Map.py` files instead of having so many additions in the `Readme.md`.

Commit:
Made .gitignore and added debugging features to beatBruteBot.py

Commit:
Added new instructions on how to edit Game.py and Map.py to support debug mode.

Commit:
Fixed mistakes in Readme.md

Celqaz commented:
Hi, @FedeCuci , 

Those are awesome features, and thanks for your contribution! 

I'll test the code with my bots and merge the PR after that.

Celqaz commented on .gitignore:
This .gitignore file is only necessary during the development process for version control; the user don't need this file to run the scripts. 

Therefore, I propose to delete this file in the upcoming merge.

FedeCuci replied:
Yes! That's my mistake in uploading it to the remote...

Celqaz replied:
Never mind. I also accidentally added the `.DS_Store` file to the repository 🤣 I'll remove mine as well.

FedeCuci replied:
😂 💯

FedeCuci replied:
Actually, I just checked and I am not sure how to remove the `.gitignore` from the pull request 😅 I'll have a look tomorrow.

FedeCuci replied:
I think I should have also deleted the `.gitignore` file from the PR. I'll close this thread now!

Celqaz commented on README.md:
````diff
@@ -76,7 +76,7 @@ def play(self):
 
 ```python
-def play(self):
+def play(self, debug):
````
By default, the `game.play()` function within `app.py` doesn't contain any other parameter; so after we changed the `Game.py`, if we still want to run `app.py`, a 'Game.play() missing 1 required positional argument: 'debug'' error will be thrown.

A possible solution maybe :
Define the `debug` parameter in the `__init__` function of Game class, with a default value of False.

```python
class Game:
	def __init__(self, player, map, maxEnergy, latency, visuals, cls, debug= False):
		// ...
		self.debug = debug

	def play(self):
		// ...
			if self.debug:
				// ...
```

FedeCuci replied:
Oh yeah my bad, I actually forgot to add this to the readme. I simply added the `debug` parameter to the `play` method in `game.py` like so: `def play(self, debug)`. 

Both solutions can work I think. The main difference I see is that having debug as parameter in each function call can be verbose and might require changes that are more spread out over files. Up to you!

Celqaz commented on README.md:
````diff
@@ -86,6 +86,67 @@ def play(self):
+
+```python
+self.currentSign = self.map[self.botPosition[0]][self.botPosition[1]]
+self.map[self.botPosition[0]][self.botPosition[1]] = 'B'
````
I tested the scripts using RandomBot, and it's a bit challenging to distinguish my robot ('B') from several '8's. Therefore, I suggest we continue using '\xC6(Æ)' to denote the robot.

```python
self.map[self.botPosition[0]][self.botPosition[1]] = '\xC6'
```

FedeCuci replied:
Yeah that's completely fine, I changed it to 'B' because I accidentally deleted '\xC6(Æ)' and couldn't find it anymore haha

Celqaz commented on README.md:
````diff
@@ -86,6 +86,67 @@ def play(self):
+def moveRobot(self, move, debug=False)
+```
+
+and change the lines after:
````
This change might cause a slight efficiency issue because the subsequent code will execute regardless of whether we are in debug mode or not, even when we set `VISUALS = False` in `beatBruteBot.py` by default. Hence, we might need to include conditional statements here to ensure that the code below runs only in debug mode.

Additionally, there's a potential side effect: if we execute `app.py`, it would still alter the display of grid cells. Therefore, I recommend retaining the original code to prevent any conflicts.

A possible solution maybe:
```python
class Map:
	// ...
	def moveRobot(self, move, debug=False):
		// ...
		if (!debug):
			// ... original code
		else:
			// change the display of grid cells.
```

FedeCuci replied:
I like the change and agree on the first point. No need to execute code that is redundant.

Could you clarify the second point, though? What do you mean by altering the display of grid cells? When does this happen? I am not seeing anything wrong when I run `app.py` myself.

Celqaz replied:
> I like the change and agree on the first point. No need to execute code that is redundant.
> 
> Could you clarify the second point, though? What do you mean by altering the display of grid cells? When does this happen? I am not seeing anything wrong when I run `app.py` myself.

Like this, all the explored squares are shown as numbers rather than  the default '.' . 
Based on your experience, do you think it's necessary to keep the original style of the map when we are not in debug mode 🤔
<img width="200" alt="app_py" src="https://github.com/Celqaz/beatBruteBot/assets/11701323/cd548b30-36f0-49c4-bd9d-885009a2221a">

FedeCuci replied:
Hey, sorry, I have been away for a bit, but I think we mean two different things when we say "alter" the outline of the map. What I meant to say is that if there is a double-digit number, which takes two places, the walls to the right will also be moved to the right by one, as seen in the image below, where I put the number 10 in one of the rows. You can see that the right outer wall was moved by one, which would also affect how the bot behaves.

One-digit numbers alter the map visually but not its outline, so it is totally up to you what you prefer! I only included the number of iterations as a number because I noticed my bot would visit certain cells multiple times and wanted to figure out why, but it is definitely not necessary. Ideally, there would be a setting "Show frequency of visited cells" that the user can toggle on/off when in debug mode, but I feel like this would overcomplicate the tool, which is trying to make testing easier in the first place. And, for most people, I think keeping the default '.' would be best. I hope I understood your comment correctly :)

![image](https://github.com/Celqaz/beatBruteBot/assets/19730248/0df37a17-b38b-4d7c-b08d-65979a2fc706)

Celqaz replied:
Hi @FedeCuci ,

Sorry for the late reply; I was just caught up in some personal stuff in the past weeks. 
After the brief review, I think there are no major issues in this PR. I will double-check in the coming week, and after we finalize the code, I will merge this PR. Cheers.

Celqaz commented on README.md:
```diff
@@ -86,6 +86,67 @@ def play(self):
+			self.overlap.append('S')
+			self.map[currentPosition[0]][currentPosition[1]] = 'S'
+		else:
+			self.map[currentPosition[0]][currentPosition[1]] = self.overlap[self.overlap.index(self.currentSign) + 1]
```
Since the value of `self.botPosition` has been updated in the previous `moveRobot` function in `Map` class, `currentPosition` no longer equals to `self.botPosition`: 
- `currentPosition`: the position before the move;
- `self.botPosition`:  the latest position after the move.

Here, we are trying to use the sign of the `self.botPosition` to update the value of the `currentPosition`, which may lead to unintended behaviours:

<img width="328" alt="Possible Error_1" src="https://github.com/Celqaz/beatBruteBot/assets/11701323/0993649c-9852-414f-af17-c030ba84d334">

<img width="328" alt="Possible Error_2" src="https://github.com/Celqaz/beatBruteBot/assets/11701323/c400c6b3-7091-46ef-8d5e-b438e2c36242">

In future adjustments, we might need to use the same position variable.

FedeCuci replied:
I see, this is a good find. Is this the new `Map.py` file that Paris uploaded?

Celqaz replied:
Yea, I think I have adopted the latest version of the `Map.py`. You can run with the `RandomBot` to reproduce this issue and debug.

Celqaz commented on README.md:
```diff
@@ -86,6 +86,67 @@ def play(self):
+else:
+	if self.currentSign in self.overlap:
+		if self.currentSign == '9' or self.currentSign == 'S':
+			self.overlap.append('S')
```
If we already have a `S` in `self.overlap`, we don't need to append another one afterwards. This happened several times when I was testing with randomBot.

One possible solution may be appending 'S' to `self.overlap` right after it has been initialized. This way, the operation will only be executed once.
```python
self.overlap = [str(i) for i in range(10)]
self.overlap.append('S')
```

FedeCuci replied:
Agree!

Celqaz replied:
Changes have been committed in [82c7691](https://github.com/Celqaz/beatBruteBot/pull/1/commits/82c76910c0ebef8972a09cc1f17c8ca532bfc2b2); conversition resolved.

Celqaz reviewed (commented):
Hi @FedeCuci , 

Thank you for your contribution!

I have completed the review for the PR and left my feedback. If you have any questions or need further clarification, please feel free to reach out. 

Cheers to the New Year! 

Kind regards,
Bin Li

FedeCuci committed:
Deleted .gitignore

Celqaz committed:
Added value to `overlap` property in Map Class Initialization

These minor changes are based on our [previous discussion](https://github.com/Celqaz/beatBruteBot/pull/1/files#r1438395842)

Celqaz reviewed (changes requested):
Hi @FedeCuci , I apologize for the delay; it's been a somewhat chaotic semester that I've just managed to get through. I hope you understand.
Regarding the PR, I have completed the review and addressed the minor issues we discussed. At the moment, there is only [one remaining issue](https://github.com/Celqaz/beatBruteBot/pull/1/files#r1438389786) that needs to be fixed, and you can find instructions on how to reproduce it in my comment. Once we resolve this issue, the PR will be ready to merge. Thank you once again for your effort and contribution.
